File: InsertionSort.py
#This is the Insertion Sort Algorithm 1/23/20

def InsertionSort(array):
    
    j = 2 #By default rule of the algorithm, start at the 2nd position of the array.

    for j in range(1, len(array)):

        key = array[j] #The key is supposed to notify the next number you are looking at.

        i = j - 1 

        while i >= 0 and array[i] > key: #BEWARE, YOU ARE STILL IN THE WHILE LOOP.
            array[i + 1] = array[i]
            i = i - 1
            
            print("Processing Insertion Sort: ", array, "\n")

        print("The key is: ", key, "\n")

        array[i + 1] = key 


def main():

    #array = [5, 2, 4, 6, 1, 3]
    #array = [12, 11, 13, 5, 6]
    array = [31, 41, 59, 26, 41, 58]

    print("The array before is: ", array, "\n")

    InsertionSort(array)

    print("Post Insertion Sort: ", array, "\n")


main()

File: InsertionSort.cpp
// InsertionSort.cpp : This file contains the 'main' function. Program execution begins and ends there.

using namespace std;

#include <iostream>
#include <vector>
#include <algorithm>


void insertionSort(vector<int> &array) {

	for (int j = 1; j < array.size(); j++) {

		int key = array[j];

		int i = j - 1;

		cout << "i is: " << i << endl;

		while (i >= 0 && array[i] > key) {
			array[i + 1] = array[i];
			i -= 1;

			// Beware, you are still in the while loop

			cout << "Processing Insertion Sort: ";
			for (int i = 0; i < array.size(); i++) {
				cout << array[i] << " ";
			}

			cout << endl;

		}
		cout << "The key is: " << key << endl;
		array[i + 1] = key;
	}
}

int main() {
    //cout << "Hello World!\n";
	vector <int> array = { 31, 41, 59, 26, 41, 58 };
	
	cout << "Before insertion sort: ";

	for (int i = 0; i < array.size(); i++) {
		cout << array[i] << " ";
	}

	cout << endl;

	insertionSort(array);

	cout << "After insertion sort: ";

	for (int i = 0; i < array.size(); i++) {
		cout << array[i] << " ";
	}

	getchar();
	return 0;

}

// Run program: Ctrl + F5 or Debug > Start Without Debugging menu
// Debug program: F5 or Debug > Start Debugging menu

// Tips for Getting Started: 
//   1. Use the Solution Explorer window to add/manage files
//   2. Use the Team Explorer window to connect to source control
//   3. Use the Output window to see build output and other messages
//   4. Use the Error List window to view errors
//   5. Go to Project > Add New Item to create new code files, or Project > Add Existing Item to add existing code files to the project
//   6. In the future, to open this project again, go to File > Open > Project and select the .sln file
